layer2/lib/mtm/merkle_tree_map.ml

```diff
@@ -0,0 +1,670 @@
+      : (k, v) op * k proof * (k, v) hnode option * result
+    =
+    (* Updates a possibly missing node. 
+          If the node is empty then we try to insert. If not empty then we try to update its content *)
+    let rec update_opt_node (current_node_opt : (k, v) hnode option)
+        : (k, v) op * k proof * (k, v) hnode option * result
+      =
+      current_node_opt
+      |> OptionExt.fold_lazy ~none:handle_key_not_found ~some:update_non_empty_node
+    (* We reached a situation when the node with the given key was not found. 
+          If the updater return a value we need to create a new leaf for it. Otherwise nothing will happen *)
+    (* Some value in the subtree starting at current_node needs to be updated. 
+          Depending on the key value we will update the current node or the left subtree or the right subtree *)
+    and update_non_empty_node (current_node : (k, v) hnode)
+        : (k, v) op * k proof * (k, v) hnode option * result
+      =
+      if current_node.content.key = key
+      then handle_key_found current_node
+      else if current_node.content.key < key
+      then update_right_subtree current_node
+      else (*i.e. current_node.key > key *) update_left_subtree current_node
+    (* We found the node that needs to be updated. 
+          If the updater returns a Some new value then we replace the old value in the node with the new one.
+          If the updater returns None then we remove the current_node*)
+    (* Go right with the search.*)
+    and update_right_subtree (current_node : (k, v) hnode)
+        : (k, v) op * k proof * (k, v) hnode option * result
+      =
+      let op, update_right_proof, new_right_node, result =
+        update_opt_node current_node.content.right
+      in
+      let p =
+        Proof.GoRight
+          { lhash = H.node_hash current_node.content.left
+          ; right_proof = update_right_proof
+          ; key = current_node.content.key
+          ; vhash = current_node.content.vhash
+          }
+      in
+      let n = H.node { current_node.content with right = new_right_node } in
+      op, p, Some n, result
+    (* Go left with the search *)
+    and update_left_subtree (current_node : (k, v) hnode)
+        : (k, v) op * k proof * (k, v) hnode option * result
+      =
+      let op, update_left_proof, new_left_node, result =
+        update_opt_node current_node.content.left
+      in
+      let p =
+        Proof.GoLeft
+          { rhash = H.node_hash current_node.content.right
+          ; left_proof = update_left_proof
+          ; key = current_node.content.key
+          ; vhash = current_node.content.vhash
+          }
+      in
+      let n = H.node { current_node.content with left = new_left_node } in
+      op, p, Some n, result
+    in
+    update_opt_node root
+  ;;
+
+  (* A convenience function that wraps a process_tree handler that doesn't return anything and only updates the current node *)
+  let write_only_handler
```

Comment:
This is not really an handler, I feel like a name like `make_write_only_handler` would be clearer

Reply:
no longer valid. the whole function was removed